### PR TITLE
types: convert EOF error to ErrTruncate when parse duration

### LINKED ISSUE #10308: Lost connection to tidb-server when query a wrong format `TIME`
## Bug Report

1. What did you do?
`select time("- -")`
`select time("---1)`
`select time("-- --1)`

2. What did you expect to see?
report errors correctly, instead of lost connection to ` tidb-server`.

3. What did you see instead?

```
> select time("- -")
Reconnecting...
Reconnecting...
```
 And here is the log of `tidb-server` when I trigger the bug:
```
[2019/04/29 18:51:56.421 +08:00] [INFO] [server.go:361] ["new connection"] [conn=1] [remoteAddr=127.0.0.1:52014]
[2019/04/29 18:51:56.624 +08:00] [INFO] [server.go:361] ["new connection"] [conn=2] [remoteAddr=127.0.0.1:52018]
[2019/04/29 18:51:57.629 +08:00] [INFO] [server.go:363] ["close connection"] [conn=2]
[2019/04/29 18:52:01.997 +08:00] [INFO] [server.go:363] ["close connection"] [conn=1]
[2019/04/29 18:52:03.200 +08:00] [INFO] [server.go:361] ["new connection"] [conn=3] [remoteAddr=127.0.0.1:52022]
[2019/04/29 18:52:03.202 +08:00] [INFO] [server.go:363] ["close connection"] [conn=3]
[2019/04/29 18:52:04.219 +08:00] [INFO] [server.go:361] ["new connection"] [conn=4] [remoteAddr=127.0.0.1:52026]
```

4. What version of `TiDB` are you using?
```
Release Version: v2.1.8
Git Commit Hash: 9a2d2da372947a50a02f9b9238a49f2db7ab9971
Git Branch: HEAD
UTC Build Time: 2019-04-12 07:48:47
GoVersion: go version go1.11.2 linux/amd64
Race Enabled: false
TiKV Min Version: 2.1.0-alpha.1-ff3dd160846b7d1aed9079c389fc188f7f5ea13e
Check Table Before Drop: false
```

### PR DESCRIPTION
**Pull request by wjhuang2016**
### What problem does this PR solve? 
fix #10308 

### What is changed and how it works?
convert EOF error to ErrTruncate

### Check List 

Tests 

 - Integration test

Code changes

Side effects

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

File: expression/integration_test.go [hunk 1]
```diff
@@ -2036,6 +2036,14 @@ func (s *testIntegrationSuite) TestTimeBuiltin(c *C) {
 	result.Check(testkit.Rows("<nil>"))
 	result = tk.MustQuery(`select tidb_parse_tso(-1)`)
 	result.Check(testkit.Rows("<nil>"))
+
+	// fix issue 10308
+	result = tk.MustQuery("select time(\"- -\");")
+	result.Check(testkit.Rows("00:00:00"))
```
Comment (zz-jason) [inline expression/integration_test.go]: would mysql through a warning?

Comment (wjhuang2016) [reply expression/integration_test.go]: > would mysql through a warning?

Both MySQL and TiDB through a warning:
```
mysql> show warnings;
+---------+------+---------------------------------------+
| Level   | Code | Message                               |
+---------+------+---------------------------------------+
| Warning | 1292 | Truncated incorrect time value: '- -' |
+---------+------+---------------------------------------+
1 row in set (0.00 sec)
```

Comment (zz-jason) [reply expression/integration_test.go]: Got it. Could you also test the warning message?

Comment (wjhuang2016) [reply expression/integration_test.go]: > Got it. Could you also test the warning message?

Done

**Pull request merged**